[BUG] The Vault action fails if the secret key contains a dot when using wildcards

## Vault server version
v1.8.2 enterprise

## vault-action version
v2

## Describe the bug
The Vault Action fails if any secret key contains a dot when using wildcards
Example: `addvertise_link.json`

## To Reproduce
Fetch all secrets in certain folder
```
- name: Get Secrets from Vault
        uses: hashicorp/vault-action@v2
        id: import-secrets
        with:
          url: "https://vault...."
          method: kubernetes
          role: role
          path: path
          kubernetesTokenPath: /var/run/secrets/kubernetes.io/serviceaccount/token
          tlsSkipVerify: false
          secrets: |
            folder/data/key1/key2/key3/key4 * | CONFIG_;
```
But if there is a secret include dot when using wildcards, such as `addvertise_link.json`. The workflow will fail.

Error message
`Error: Unable to retrieve result for data.data."`*`". No match data was found. Double check your Key or Selector.`


## Expected behavior
Get all secrets in specific folder

@fairclothjm thanks for quick response. After I tested with my case, but still got the following error.
```
Run hashicorp/vault-action@VAULT-26014/wildcard-for-dot-fields-fix
node:internal/modules/cjs/loader:1048
  const err = new Error(message);
              ^

Error: Cannot find module 'jsonata'
Require stack:
- /home/runner/_work/_actions/hashicorp/vault-action/VAULT-26014/wildcard-for-dot-fields-fix/dist/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Module._load (node:internal/modules/cjs/loader:901:27)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:130:18)
    at file:///home/runner/_work/_actions/hashicorp/vault-action/VAULT-26014/wildcard-for-dot-fields-fix/dist/index.js:1[42](https://github.com/Houzz/c2/actions/runs/8682786252/job/23807820689#step:2:43)70:17
    at file:///home/runner/_work/_actions/hashicorp/vault-action/VAULT-26014/wildcard-for-dot-fields-fix/dist/index.js:1[46](https://github.com/Houzz/c2/actions/runs/8682786252/job/23807820689#step:2:47)82:3
    at ModuleJob.run (node:internal/modules/esm/module_job:217:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:316:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:66:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/runner/_work/_actions/hashicorp/vault-action/VAULT-26014/wildcard-for-dot-fields-fix/dist/index.js'
  ]
}

Node.js v20.8.1
```

the key value in my test case is
key: 
`key.json`
value:
`{
  "test": {
    "key1": "123"
  },
  "test2": {
    "key2": "456"
  }
}`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 ## Unreleased
 
+Improvements:
+
+* fix wildcard handling when field contains dot [GH-542](https://github.com/hashicorp/vault-action/pull/542)
+
 Features:
 
 * `secretId` is no longer required for approle to support advanced use cases like machine login when `bind_secret_id` is false. [GH-522](https://github.com/hashicorp/vault-action/pull/522)

--- a/integrationTests/basic/integration.test.js
+++ b/integrationTests/basic/integration.test.js
@@ -31,6 +31,14 @@ describe('integration', () => {
             },
         });
 
+        await got(`${vaultUrl}/v1/secret/data/test-with-dot-char`, {
+            method: 'POST',
+            headers: {
+                'X-Vault-Token': vaultToken,
+            },
+            body: `{"data":{"secret.foo":"SUPERSECRET"}}`
+        });
+
         await got(`${vaultUrl}/v1/secret/data/nested/test`, {
             method: 'POST',
             headers: {
@@ -191,6 +199,16 @@ describe('integration', () => {
         expect(core.exportVariable).toBeCalledWith('SECRET', 'SUPERSECRET');
         expect(core.exportVariable).toBeCalledWith('NAMED_SECRET', 'SUPERSECRET');
         expect(core.exportVariable).toBeCalledWith('OTHERSECRETDASH', 'OTHERSUPERSECRET');
+    });
+
+    it('get wildcard secrets with dot char', async () => {
+        mockInput(`secret/data/test-with-dot-char * ;`);
+
+        await exportSecrets();
+
+        expect(core.exportVariable).toBeCalledTimes(1);
+
+        expect(core.exportVariable).toBeCalledWith('SECRET__FOO', 'SUPERSECRET');
     });
 
     it('get wildcard secrets', async () => {

--- a/src/secrets.js
+++ b/src/secrets.js
@@ -59,7 +59,7 @@ async function getSecrets(secretRequests, client, ignoreNotFound) {
 
         body = JSON.parse(body);
 
-        if (selector == WILDCARD) {                     
+        if (selector == WILDCARD) {
             let keys = body.data;
             if (body.data["data"] != undefined) {
                 keys = keys.data;
@@ -67,20 +67,26 @@ async function getSecrets(secretRequests, client, ignoreNotFound) {
 
             for (let key in keys) {
                 let newRequest = Object.assign({},secretRequest);
-                newRequest.selector = key;                  
-                
+                newRequest.selector = key;
+
                 if (secretRequest.selector === secretRequest.outputVarName) {
                     newRequest.outputVarName = key;
-                    newRequest.envVarName = key;        		
-                }
-                else {
+                    newRequest.envVarName = key;
+                } else {
                     newRequest.outputVarName = secretRequest.outputVarName+key;
-                    newRequest.envVarName = secretRequest.envVarName+key;        		
+                    newRequest.envVarName = secretRequest.envVarName+key;
                 }
 
                 newRequest.outputVarName = normalizeOutputKey(newRequest.outputVarName);
-                newRequest.envVarName = normalizeOutputKey(newRequest.envVarName,true);   
+                newRequest.envVarName = normalizeOutputKey(newRequest.envVarName,true);
 
+                // JSONata field references containing reserved tokens should
+                // be enclosed in backticks
+                // https://docs.jsonata.org/simple#examples
+                if (key.includes(".")) {
+                    const backtick = '`';
+                    key = backtick.concat(key, backtick);
+                }
                 selector = key;
 
                 results = await selectAndAppendResults(
@@ -94,13 +100,13 @@ async function getSecrets(secretRequests, client, ignoreNotFound) {
         }
         else {
           results = await selectAndAppendResults(
-            selector, 
-            body, 
-            cachedResponse, 
-            secretRequest, 
+            selector,
+            body,
+            cachedResponse,
+            secretRequest,
             results
           );
-        }   
+        }
     }
 
     return results;

